appengine/genetics/js/cage.js

```diff
@@ -0,0 +1,725 @@
+  var opponent = Genetics.Cage.miceMap_[chosen.id];
+  // Check if opponent is itself.
+  if (mouse.id == opponent.id) {
+    new Genetics.Cage.Event('FIGHT', mouse.id, 'SELF').addToQueue();
+    Genetics.Cage.death(mouse);
+    return;
+  }
+  // Compute winner using size to weigh probability of winning.
+  var fightResult = Math.random() - mouse.size / (mouse.size + opponent.size);
+  if (fightResult < 0) {  // If fight proposer won.
+    new Genetics.Cage.Event('FIGHT', mouse.id, 'WIN', opponent.id).addToQueue();
+    Genetics.Cage.death(opponent);
+  }
+  else if (fightResult == 0) {  // If it was a tie.
+    new Genetics.Cage.Event('FIGHT', mouse.id, 'TIE', opponent.id).addToQueue();
+  } else {  // If opponent won.
+    new Genetics.Cage.Event('FIGHT', mouse.id, 'LOSS', opponent.id)
+        .addToQueue();
+    Genetics.Cage.death(mouse);
+  }
+};
+
+/**
+ * Gets requested mate of mouse, attempts mating and starts life of any
+ * resultant child.
+ * @param {Genetics.Mouse} mouse The mouse to initiate mating.
+ */
+Genetics.Cage.tryMate = function(mouse) {
+  var result = Genetics.Cage.runMouseFunction(mouse, 'chooseMate');
+  // Check if function threw an error or caused a timeout.
+  if (result[0] == 'FAILURE') {
+    return;
+  }
+  var chosen = result[1];
+  // Check if no mouse was chosen.
+  if (chosen === null) {
+    new Genetics.Cage.Event('MATE', mouse.id, 'NONE').addToQueue();
+    mouse.fertility = 0;
+    return;
+  }
+  // Check if return value is invalid
+  if (typeof chosen != 'object' ||
+      Genetics.Cage.miceMap_[chosen.id] == undefined) {
+    new Genetics.Cage.Event('MATE', mouse.id, 'INVALID').addToQueue();
+    mouse.fertility = 0;
+    return;
+  }
+  // Retrieve local copy of mate to ensure that mouse was not modified.
+  var mate = Genetics.Cage.miceMap_[chosen.id];
+  if (Genetics.Cage.isMatingSuccessful(mouse, mate)) {
+    // Create offspring from the two mice.
+    Genetics.Cage.createOffspring(mouse, mate);
+    // Account for overpopulation.
+    if (Genetics.Cage.aliveMice_.length > Genetics.Cage.MAX_POPULATION) {
+      // Find oldest mouse.
+      var oldestMouse = Genetics.Cage.aliveMice_[0];
+      for (var i = 1; i < Genetics.Cage.aliveMice_.length; i++) {
+        var aliveMouse = Genetics.Cage.aliveMice_[i];
+        if (!oldestMouse || aliveMouse.age > oldestMouse.age) {
+          oldestMouse = aliveMouse;
+        }
+      }
+      // Kill oldest mouse.
+      new Genetics.Cage.Event('OVERPOPULATION', oldestMouse.id).addToQueue();
+      Genetics.Cage.death(oldestMouse);
+    }
+  } else {
+    // Use up one mating attempt for mouse.
+    mouse.fertility--;
+  }
+};
+
+/**
+ * Creates a child mouse from two mice and adds offspring to the cage.
+ * @param {!Genetics.Mouse} parent1 One of the parents of the new mouse.
+ * @param {!Genetics.Mouse} parent2 One of the parents of the new mouse.
+ */
+Genetics.Cage.createOffspring = function(parent1, parent2) {
```

Comment:
The function would still work if the parents are the same, but in the game it is not desired. There is a check for this in tryMate.